libraries/AP_RangeFinder/AP_RangeFinder_Ainstein_LR_D1.cpp

```diff
@@ -0,0 +1,154 @@
+
+bool AP_RangeFinder_Ainstein_LR_D1::move_signature_in_buffer(uint8_t start)
+{
+    for (uint8_t i=start; i<buffer_used; i++) {
+        if (u.buffer[i] == PACKET_HEADER_MSB) {
+            memmove(&u.buffer[0], &u.buffer[i], buffer_used-i);
+            buffer_used -= i;
+            return true;
+        }
+    }
+    // header byte not in buffer
+    buffer_used = 0;
+    return false;
+}
+
+uint8_t AP_RangeFinder_Ainstein_LR_D1::LRD1Union::calculate_checksum() const
+{
+    uint32_t sum = 0;
+    for (uint8_t i=3; i<31; i++) {
+        sum += buffer[i];
+    }
+    return sum & 0xff;
+}
+
+// get_reading - read a value from the sensor
+bool AP_RangeFinder_Ainstein_LR_D1::get_reading(float &reading_m)
+{
+    if (uart == nullptr) {
+        return false;
+    }
+
+    const uint8_t num_read = uart->read(&u.buffer[buffer_used], ARRAY_SIZE(u.buffer)-buffer_used);
+    buffer_used += num_read;
+
+    if (buffer_used == 0) {
+        return false;
+    }
+
+    if (u.packet.header_msb != PACKET_HEADER_MSB &&
+        !move_signature_in_buffer(1)) {
+        return false;
+    }
+
+    if (buffer_used < sizeof(u.packet)) {
+        return false;
+    }
+
+    // sanity checks; see data sheet on these fixed values.
+    if (u.packet.header_lsb != PACKET_HEADER_LSB ||
+        u.packet.device_id != 0x00 ||
+        u.packet.length != 28 ||
+        u.calculate_checksum() != u.packet.checksum) {
+        // sanity checks failed - discard and try again next time we're called:
+        move_signature_in_buffer(1);
+        return false;
+    }
+
+    reading_m = be16toh(u.packet.object1_alt) * 0.01;
+
+    // consume this packet:
+    move_signature_in_buffer(sizeof(u.packet));
+
+    /*  From datasheet:
+      Altitude data from radar may have unexpected
+      or incorrect readings if aircraft pitch and/or
+      roll are beyond the radar’s detection angle
+      (Azimuth 43°，Elevation 30°). The ideal case is
+      keeping radar perpendicular to ground.
+    */
+#ifndef HAL_BUILD_AP_PERIPH   // we may need to move this elsewhere to the entire function or driver?
+    const auto &ahrs = AP::ahrs();
+    if (fabs(ahrs.roll) > radians(43)) {
+        return false;
+    }
+    if (fabs(ahrs.pitch) > radians(30)) {
+        return false;
+    }
+#endif
+
+    snr = u.packet.object1_snr;
+
+    /* From datasheet:
+        Altitude measurements associated with a SNR value 
+        of 13dB or lower are considered erroneous
+    */
+    if (snr <= 13)
+    {
+        signal_quality_pct = RangeFinder::SIGNAL_QUALITY_MIN;
+        // out of range according to datasheet
+        // max valid altitude from device is 655 metres
+        if (snr == 0)
+        {
+            reading_m = MAX(656, max_distance_cm() * 0.01 + 1);
+            return true;
+        }
+        
+        return false;
+    }
+
+    signal_quality_pct = RangeFinder::SIGNAL_QUALITY_MAX;
+
+    check_for_malfunction();
```

Comment:
The way this is written not, this has the potential of spamming the GCS. Needs a throttle of some sort to not spam any faster than 1Hz or just once (for each malfunction) by caching the value and only printing on malfunction change.. Don't forget to nerf signal_quality_pct when it's in an error state